Add Webhook debugging guide, updates to sync guide (#851)
clerk/clerk-docs: docs/integrations/webhooks/debug-your-webhooks.mdx

## Webhooks and local development

When you or a user of your application performs certain actions, a webhook can be triggered. You can see the full list of [webhook events](/docs/integrations/webhooks/overview#supported-webhook-events) for a list of the actions that could result in a Webhook. Depending on the events subscribed to in the [Webhook section](https://dashboard.clerk.com/last-active?path=webhooks) of your dashboard, a webhook event will be triggered and sent to the specified endpoint in your application.

When you are developing on your localhost, your application is not internet facing and can't receive the webhook request. You will need to use a tool that creates a tunnel from the internet to your localhost. These tools will provide temporary or permanant URLs depending on the tool and the plan you subscribe to. Popular tools include `ngrok`, `localtunnel` and `Cloudflare Tunnel`.

Below is an image demonstrating the flow from when you perform an action that triggers a webhook through until the webhook is received by your application. This flow chart will show the role of the tunnel.

> FLOW CHART IMAGE HERE

Reviewer: Missing image

Author: I sent you this image before via DMs for feedback/etc.

With the Webhook flow in mind, the following areas could cause problems when trying to develop with webhooks:

> 1. Middleware is not configured correctly for webhooks

Reviewer: Link these options to their corresponding sections, so `[Middleware is not configured correctly for webhooks](#check-your-middleware-configuration)`

2. The webhook route is not correctly configured
3. Configuration in your dashboard
4. The tunnel configuration
5. Verification of the webhook
6. The logic you created to process the webhook

> ##  Check your Middleware configuration

Reviewer: Update for clerkMiddleware

Incoming webhook events will never be signed in -- they are coming from a source outside of your application. Since they will be in a signed out state, the simplest configuration is to use `ignoredRoutes` for your webhook routes. Below is an example of making all webhook routes ignored.

```tsx filename="middleware.tsx" {4}
import { authMiddleware } from "@clerk/nextjs";

export default authMiddleware({
  ignoredRoutes: ["/api/webhooks(.*)"],
});

export const config = {
  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

## Test the Route Handler or API Route

If you are having trouble with your webhook, you can create a basic Route Handler and then test it locally.  First, add the following file and code to your app.

```ts filename="app/webhooks/test/route.ts"
export async function POST() {
   return Response.json({ message: 'The route is working' });
}
```

While your application is running, try this from the command line.

```bash
curl -H 'Content-Type: application/json' \
     -X POST http://localhost:3000/api/webhooks/test
```

If you see the `{"message":"The route is working"}` then the basic Route Handler is working, and ready to build on.

> ## Check your dashboard configuration

Reviewer: If we aren't including mentioning the "Clerk Dashboard" explicitly anywhere in this section, then we should at least put it in the title so users know we are talking about the Clerk Dashboard

```suggestion
## Check the webhook configuration in the Clerk Dashboard
```

| 400  | This error usually indicates the verification failed, but could be caused by other issues |
| 401  | The request was not authorized. If your test in 'Test the Route Handler or API Route' worked, you should not see this error. If you are seeing it, then you will need to configure your Middleware to accept the request, |
| 404  | The URL for the webhook was not found. This indicates that either the URL you added to the Webhook in the dashboard was incorrect, your application is not running, or the path is wrong.|
| 405  | Your route is not accepting a POST request. All webhooks are POST requests, to the route must accept them. Unless you are using the route for somethinh else, you can restrict the route to POST requests only. |
| 500  | A 500 error indicates that something failed in the webhook code. The request made it to your application, but there is a code related issue. This is likely a webhook verification issue or something in the webhook logic. Please see the following sections.

You may run into other errors. See [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) from MDN for reference.

## Debug webhook verfication

> To verify the webhook, your webhook route will need the Signing Secret from the Webhook section of the dashboard.  It will also need the Svix headers which are included in the request. Please see [Create the endpoint in your application](/docs/integrations/webhooks/sync-data#create-the-endpoint-in-your-application) for a detailed code example. You can also visit the [guide](https://docs.svix.com/receiving/verifying-payloads/how) on verifying webhooks with the Svix library.

Reviewer: ```suggestion
To verify the webhook, your webhook route will need the **Signing Secret** from the Clerk Dashboard and the Svix headers from the payload. See the [Create a webhook](/docs/integrations/webhooks/create-a-webhook#create-a-webhook-handler) guide for a detailed code example. You can also visit the [Svix guide](https://docs.svix.com/receiving/verifying-payloads/how) on verifying webhooks.
```

When you are developing on your localhost, your application is not internet facing and can't receive the webhook request. You will need to use a tool that creates a tunnel from the internet to your localhost. These tools will provide temporary or permanant URLs depending on the tool and the plan you subscribe to. Popular tools include `ngrok`, `localtunnel` and `Cloudflare Tunnel`.

With the Webhook flow in mind, the following areas could cause problems when trying to develop with webhooks:

1. [Middleware is not configured correctly for webhooks](#check-your-middleware-configuration)
2. [The webhook route is not correctly configured](#test-the-route-handler-or-api-route)
3. [Configuration in your dashboard](#check-your-configuration-in-the-clerk-dashboard)
4. [The tunnel configuration](#debug-your-tunnel-and-webhook-delivery)
5. [Verification of the webhook](#debug-webhook-verification)
6. [The logic you created to process the webhook](#check-your-logic)

##  Check your Middleware configuration

> Incoming webhook events will never be signed in -- they are coming from a source outside of your application. Since they will be in a signed out state, the route should be public. The simplest configuration is to use `ignoredRoutes` for your webhook routes when using `authMiddleware` and applying no protection to the route with `clerkMiddleware`. Below is an example of making all webhook routes ignored.

Reviewer: ```suggestion
Incoming webhook events will never be signed in -- they are coming from a source outside of your application. Since they will be in a signed out state, the route should be public. 

The following example shows the recommended Middleware configuration for your webhook routes.
```
